From 48b46c65087f38fe941dc925bd5cf8cea2dc6d21 Mon Sep 17 00:00:00 2001
From: Michal Hrusecky <michal.hrusecky@turris.com>
Date: Thu, 6 May 2021 22:25:03 +0200
Subject: [PATCH] alert_luajit: Make MAC addresses available

Add ether_src and ether_dst to the packet information in LUA to allow
better alerting.
---
 src/loggers/alert_luajit.cc    | 30 ++++++++++++++++++++++++++++++
 src/managers/lua_plugin_defs.h |  2 ++
 src/protocols/eth.h            |  1 +
 3 files changed, 33 insertions(+)

diff --git a/src/loggers/alert_luajit.cc b/src/loggers/alert_luajit.cc
index 0783e2fd5..a6fb19d96 100644
--- a/src/loggers/alert_luajit.cc
+++ b/src/loggers/alert_luajit.cc
@@ -21,6 +21,7 @@
 #include "config.h"
 #endif
 
+#include "protocols/eth.h"
 #include "detection/ips_context.h"
 #include "detection/signature.h"
 #include "events/event.h"
@@ -82,6 +83,35 @@ SO_PUBLIC const SnortPacket* get_packet()
     lua_packet.sp = packet->ptrs.sp;
     lua_packet.dp = packet->ptrs.dp;
 
+    if ( !(packet->proto_bits & PROTO_BIT__ETH) ) {
+        lua_packet.ether_dst = lua_packet.ether_src = "";
+        return &lua_packet;
+    }
+
+    static THREAD_LOCAL char eth_src[eth::ETH_ADDR_STR_LEN], eth_dst[eth::ETH_ADDR_STR_LEN];
+
+    const eth::EtherHdr* eh = layer::get_eth_layer( packet );
+
+    if(eh && eh->ether_src) {
+        snprintf(eth_src, eth::ETH_ADDR_STR_LEN, "%02X:%02X:%02X:%02X:%02X:%02X", eh->ether_src[0],
+            eh->ether_src[1], eh->ether_src[2], eh->ether_src[3],
+            eh->ether_src[4], eh->ether_src[5]);
+
+        lua_packet.ether_src = (const char*)eth_src;
+    } else {
+        lua_packet.ether_src = "";
+    }
+
+    if(eh && eh->ether_dst) {
+        snprintf(eth_dst, eth::ETH_ADDR_STR_LEN, "%02X:%02X:%02X:%02X:%02X:%02X", eh->ether_dst[0],
+            eh->ether_dst[1], eh->ether_dst[2], eh->ether_dst[3],
+            eh->ether_dst[4], eh->ether_dst[5]);
+
+        lua_packet.ether_dst = (const char*)eth_dst;
+    } else {
+        lua_packet.ether_dst = "";
+    }
+
     return &lua_packet;
 }
 
diff --git a/src/managers/lua_plugin_defs.h b/src/managers/lua_plugin_defs.h
index 4efc683dd..5fce3682e 100644
--- a/src/managers/lua_plugin_defs.h
+++ b/src/managers/lua_plugin_defs.h
@@ -55,6 +55,8 @@ struct SnortPacket
     uint64_t num;
     unsigned sp;
     unsigned dp;
+    const char* ether_src;
+    const char* ether_dst;
 };
 
 extern "C"
diff --git a/src/protocols/eth.h b/src/protocols/eth.h
index 5565468f0..d87ecf146 100644
--- a/src/protocols/eth.h
+++ b/src/protocols/eth.h
@@ -33,6 +33,7 @@ namespace eth
 constexpr uint16_t MTU_LEN = 1500;
 constexpr uint16_t MAX_FRAME_LENGTH = 1500;
 constexpr uint16_t ETH_HEADER_LEN = 14;
+constexpr uint16_t ETH_ADDR_STR_LEN = 18;
 
 struct EtherHdr
 {
